fix(table): avoid panic when matching equality-delete partitions

### Apache Iceberg version

main (development)

### Please describe the bug 🐞

# Equality-delete scan planning can panic when partition values are non-comparable

## Summary

`table/scanner.go` performs direct `!=` comparisons on partition values during equality-delete scan planning. While this works for comparable Go types, it panics when partition values are non-comparable (for example, slices).

As a result, scan planning can crash instead of returning a normal error.

---

## Impact

- Equality-delete scan planning can panic unexpectedly.
- Any scan path that calls `partitionsMatch` is affected.
- Causes a runtime panic (`comparing uncomparable type`) instead of graceful error handling.
- This is a correctness and stability issue.

---

## Affected Location

```
table/scanner.go
Lines: 547-558
```

---

## Root Cause

The implementation compares partition values using Go's `!=` operator.

Example:

```go
if left != right {
    ...
}
```

However, Go only allows `==` and `!=` on **comparable** types.

Examples of non-comparable types include:

- `[]T` (slices)
- `map[K]V`
- functions

Attempting to compare these values results in a runtime panic:

```
panic: runtime error: comparing uncomparable type []string
```

---

## Expected Behavior

Partition matching should safely handle every supported partition value type.

For non-comparable values, the scanner should:

- use an appropriate typed comparison, or
- detect unsupported/non-comparable values and return a normal error.

Scan planning should never panic because of partition value comparison.

---

## Actual Behavior

When a partition value contains a non-comparable Go type, equality-delete planning crashes with a runtime panic similar to:

```
panic: runtime error: comparing uncomparable type
```

instead of returning an error.

---

--- a/table/scanner.go
+++ b/table/scanner.go
@@ -22,7 +22,9 @@ import (
 	"context"
 	"fmt"
 	"iter"
+	"maps"
 	"math"
+	"reflect"
 	"slices"
 	"strings"
 	"sync"
@@ -589,17 +591,7 @@ func matchEqualityDeletesToData(dataEntry iceberg.ManifestEntry, eqDeleteEntries
 }
 
 func partitionsMatch(a, b map[int]any) bool {
-	if len(a) != len(b) {
-		return false
-	}
-
-	for k, v := range a {
-		if bv, ok := b[k]; !ok || bv != v {
-			return false
-		}
-	}
-
-	return true
+	return maps.EqualFunc(a, b, reflect.DeepEqual)
 }
 
 // buildDVIndex indexes deletion vectors by the data file path they reference.

--- a/table/scanner_internal_test.go
+++ b/table/scanner_internal_test.go
@@ -53,6 +53,78 @@ func newDeleteManifest(minSeqNum int64) iceberg.ManifestFile {
 		Build()
 }
 
+func TestPartitionsMatchHandlesBinaryValues(t *testing.T) {
+	tests := []struct {
+		name  string
+		left  map[int]any
+		right map[int]any
+		match bool
+	}{
+		{
+			name:  "equal binary values",
+			left:  map[int]any{1000: []byte{0xde, 0xad}},
+			right: map[int]any{1000: append([]byte(nil), 0xde, 0xad)},
+			match: true,
+		},
+		{
+			name:  "different binary values",
+			left:  map[int]any{1000: []byte{0xde, 0xad}},
+			right: map[int]any{1000: []byte{0xbe, 0xef}},
+		},
+		{
+			name:  "binary and string values differ",
+			left:  map[int]any{1000: []byte("value")},
+			right: map[int]any{1000: "value"},
+		},
+		{
+			name:  "comparable values still match",
+			left:  map[int]any{1000: int32(7), 1001: "region"},
+			right: map[int]any{1000: int32(7), 1001: "region"},
+			match: true,
+		},
+		{
+			name:  "different field IDs",
+			left:  map[int]any{1000: []byte{1}},
+			right: map[int]any{1001: []byte{1}},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			assert.NotPanics(t, func() {
+				assert.Equal(t, tt.match, partitionsMatch(tt.left, tt.right))
+			})
+		})
+	}
+}
+
+func TestMatchEqualityDeletesToDataHandlesBinaryPartitions(t *testing.T) {
+	dataSeqNum := int64(1)
+	deleteSeqNum := int64(2)
+	dataFile := &mockDataFile{
+		path:      "data.parquet",
+		partition: map[int]any{1000: []byte{0xde, 0xad}},
+	}
+	matchingDelete := &mockDataFile{
+		path:      "matching-delete.parquet",
+		partition: map[int]any{1000: append([]byte(nil), 0xde, 0xad)},
+	}
+	nonMatchingDelete := &mockDataFile{
+		path:      "non-matching-delete.parquet",
+		partition: map[int]any{1000: []byte{0xbe, 0xef}},
+	}
+
+	dataEntry := iceberg.NewManifestEntry(
+		iceberg.EntryStatusADDED, nil, &dataSeqNum, nil, dataFile)
+	deleteEntries := []iceberg.ManifestEntry{
+		iceberg.NewManifestEntry(iceberg.EntryStatusADDED, nil, &deleteSeqNum, nil, nonMatchingDelete),
+		iceberg.NewManifestEntry(iceberg.EntryStatusADDED, nil, &deleteSeqNum, nil, matchingDelete),
+	}
+
+	assert.Equal(t, []iceberg.DataFile{matchingDelete},
+		matchEqualityDeletesToData(dataEntry, deleteEntries))
+}
+
 func TestMinSequenceNum(t *testing.T) {
 	tests := []struct {
 		name      string

